nltk/tokenize/casual.py

```
@@ -0,0 +1,304 @@
+
+WORD_RE = re.compile(r"""(%s)""" % "|".join(REGEXPS), re.VERBOSE | re.I
+                     | re.UNICODE)
+
+# The emoticon string gets its own regex so that we can preserve case for
+# them as needed:
+EMOTICON_RE = re.compile(EMOTICONS, re.VERBOSE | re.I | re.UNICODE)
+
+# These are for regularizing HTML entities to Unicode:
+HTML_ENTITY_DIGIT_RE = re.compile(r"&#\d+;")
+HTML_ENTITY_ALPHA_RE = re.compile(r"&\w+;")
+AMP = "&amp;"
+
+######################################################################
+
+class TweetTokenizer:
+    """Tokenize Tweets"""
+    def __init__(self, preserve_case=True):
+        self.preserve_case = preserve_case
+
+    def tokenize(self, text):
+        """
+        :param s: str
+        :rtype: list(str)
+        :return: a tokenized list of strings; concatenating this list returns
+        the original string if preserve_case=False
+        """
+        # Try to ensure unicode:
+        try:
+            text = str(text)
+        except UnicodeDecodeError:
+            text = str(text).encode('string_escape')
+            text = str(text)
+        # Fix HTML character entities:
+        text = self._html2unicode(text)
+        # Tokenize:
+        words = WORD_RE.findall(text)
+        # Possibly alter the case, but avoid changing emoticons like :D into :d:
+        if not self.preserve_case:
+            words = list(map((lambda x : x if EMOTICON_RE.search(x) else
+                              x.lower()), words))
+        return words
+
+
+    def _html2unicode(self, text):
+        """
+        Try to replace all the HTML entities in `s` with their corresponding
+        Unicode characters.
+        """
+        # First the digits:
+        ents = set(HTML_ENTITY_DIGIT_RE.findall(text))
+        if len(ents) > 0:
+            for ent in ents:
+                entnum = ent[2:-1]
+                try:
+                    entnum = int(entnum)
+                    text = text.replace(ent, chr(entnum))
+                except:
```

It'd be ValueError for int(..) and chr(..) calls, but if the idea is to ignore all errors we can use `except Exception`.

I wonder if it is better to copy-paste an existing battle-tested function (like https://github.com/scrapy/w3lib/blob/master/w3lib/html.py#L38) or use routines from the standard library (http://stackoverflow.com/a/2087433/114795).